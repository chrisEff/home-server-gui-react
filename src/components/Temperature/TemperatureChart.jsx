'use strict'

import React from 'react'
import PropTypes from 'prop-types'

import { CartesianGrid, Line, LineChart, Tooltip, XAxis, YAxis } from 'recharts'

const TemperatureChart = ({ data }) => (
	<LineChart syncId="temp" width={900} height={240} data={data} margin={{ top: 5, right: 5, left: -5, bottom: 5 }}>
		<CartesianGrid strokeDasharray="2 6" stroke="rgb(135, 147, 155)" />
		<XAxis dataKey="time" ticks={['00:00', '03:00', '06:00', '09:00', '12:00', '15:00', '18:00', '21:00', '24:00']} />
		<YAxis />
		<Tooltip />
		<Line name="yesterday" dot={false} type="linear" dataKey="yesterday" stroke="blue" />
		<Line name="today" dot={false} type="linear" dataKey="today" stroke="red" />
	</LineChart>
)

TemperatureChart.propTypes = {
	data: PropTypes.array,
}

export default TemperatureChart
