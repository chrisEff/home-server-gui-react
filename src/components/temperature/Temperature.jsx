'use strict'

const api = require('../../../config').api
api.url = `${api.protocol}://${api.host}:${api.port}`

const React = require('react')
const PropTypes = require('prop-types')
const dateFormat = require('dateformat')
const sortBy = require('lodash.sortby')

const TemperatureSensor = require('./TemperatureSensor')
const ErrorMessage = require('../../ErrorMessage')

class Temperature extends React.Component {

	static propTypes = {
		title: PropTypes.string,
	}

	constructor (props) {
		super(props)
		this.state = {
			tempSensors: [],
			errorMsg: null,
		}
	}

	async componentDidMount () {
		try {
			const response = await fetch(`${api.url}/tempSensors/?key=${api.key}`)

			const date = new Date()
			date.setHours(0, 0, 0, 0)
			const timestampStartOfToday = Math.floor(date.getTime() / 1000)

			date.setDate(date.getDate() - 1)
			const timestampStartOfYesterday = Math.floor(date.getTime() / 1000)

			const tempSensors = await Promise.all(Object.values(await response.json()).map(async sensor => {
				const history = {}

				const historyToday     = (await Temperature.fetchHistory(sensor.id, timestampStartOfToday))
				const historyYesterday = (await Temperature.fetchHistory(sensor.id, timestampStartOfYesterday, timestampStartOfToday-1))

				for (let entry of historyYesterday) {
					if (entry.time % 120 === 0) {
						const time = dateFormat(new Date(entry.time * 1000), 'HH:MM')
						history[time] = { time: time, yesterday: entry.val }
					}
				}
				for (let entry of historyToday) {
					const time = dateFormat(new Date(entry.time * 1000), 'HH:MM')
					if (!history[time]) {
						history[time] = { time }
					}
					history[time].today = entry.val
				}

				sensor.history = sortBy(Object.values(history), ['time'])
				sensor.history.push({ time: '24:00' })

				return sensor
			}))
			this.setState({ tempSensors })
		} catch (e) {
			this.setState({ errorMsg: e.message })
		}
	}

	static async fetchHistory (sensorId, min, max) {
		let url = `${api.url}/tempSensors/${sensorId}/history?key=${api.key}&min=${min}`
		if (max) {
			url += `&max=${max}`
		}
		return (await fetch(url)).json()
	}

	render () {
		if (!this.state.tempSensors) {
			return <div/>
		}

		return (
			<div>
				<h2>{this.props.title}</h2>
				{this.state.errorMsg && <ErrorMessage message={this.state.errorMsg}/>}
				{this.state.tempSensors.map(sensor => <TemperatureSensor key={sensor.id} name={sensor.name} value={sensor.celsiusValue} history={sensor.history}/>)}
			</div>
		)
	}
}

module.exports = Temperature
