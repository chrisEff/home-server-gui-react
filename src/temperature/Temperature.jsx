'use strict';
const config = require('../../config')

const TemperatureSensor = require('./TemperatureSensor')

class Temperature extends React.Component {
  constructor(props) {
    super(props)
    this.state = {}
  }

  async componentDidMount() {
    const response = await fetch(`${config.api.protocol}://${config.api.host}:${config.api.port}/tempSensors/?key=${config.api.key}`)
    const tempSensors = Object.values(await response.json())
    this.setState({
      tempSensors
    })
  }

  render() {
    if (!this.state.tempSensors) {
      return <div/>
    }

    return (
      <div>
        <h2>Temperatur</h2>
        <table>
          <tbody>
            {this.state.tempSensors.map(sensor => <TemperatureSensor key={sensor.id} name={sensor.name} value={sensor.celsiusValue}/>)}
          </tbody>
        </table>
      </div>
    )
  }
}

module.exports = Temperature
