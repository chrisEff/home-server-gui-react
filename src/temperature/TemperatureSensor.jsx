'use strict'

const React = require('react')
const PropTypes = require('prop-types')

const TemperatureSensor = ({name, value}) => <tr>
	<td>{name}:</td>
	<td>{value}Â°C</td>
</tr>

TemperatureSensor.propTypes = {
	name: PropTypes.string.isRequired,
	value: PropTypes.number
}

module.exports = TemperatureSensor
