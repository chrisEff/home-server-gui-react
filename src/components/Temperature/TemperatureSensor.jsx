'use strict'

import React from 'react'
import PropTypes from 'prop-types'

import TemperatureChart from './TemperatureChart'

const TemperatureSensor = ({name, value, history}) => (
	<div className='tempSensor'>
		<h4 className='name'>{name}</h4>
		<div className='value'>{Math.round(value * 10) / 10}Â°C</div>

		{ history && <TemperatureChart data={history} /> }
	</div>
)

TemperatureSensor.propTypes = {
	name: PropTypes.string.isRequired,
	value: PropTypes.number,
	history: PropTypes.array,
}

export default TemperatureSensor
