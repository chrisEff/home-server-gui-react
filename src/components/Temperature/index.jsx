'use strict'

import React from 'react'
import PropTypes from 'prop-types'
import {connect} from 'react-redux'

import TemperatureSensor from './TemperatureSensor'

import {loadTempSensors} from '../../actions/temperature'

class Temperature extends React.Component {

	static propTypes = {
		title: PropTypes.string,
		tempSensors: PropTypes.arrayOf(
			PropTypes.object
		),
	}

	async componentDidMount () {
		this.props.onLoad()
	}

	render () {
		return !this.props.tempSensors ? <div/> : (
			<div id='temperature'>
				<h2>{this.props.title}</h2>
				{this.props.tempSensors.map(sensor =>
					<TemperatureSensor key={sensor.id} name={sensor.name} value={sensor.celsiusValue} history={sensor.history}/>
				)}
			</div>
		)
	}
}

const mapStateToProps = state => ({
	tempSensors: state.tempSensors,
})

const mapDispatchToProps = dispatch => ({
	onLoad: () => dispatch(loadTempSensors()),
})

export default connect(mapStateToProps, mapDispatchToProps)(Temperature)
