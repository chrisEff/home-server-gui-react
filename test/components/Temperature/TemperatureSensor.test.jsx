'use strict'

import React from 'react'
import {TemperatureSensor} from '../../../src/components/Temperature/TemperatureSensor'

import {shallow} from 'enzyme'

describe('TemperatureSensor', () => {
	it('should render correctly', () => {
		const wrapper = shallow(<TemperatureSensor sensor={{name: 'Office', celsiusValue: 23.187}} />)
		expect(wrapper).toMatchSnapshot()
	})
})
