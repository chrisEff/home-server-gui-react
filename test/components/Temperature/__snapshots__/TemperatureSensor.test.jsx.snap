// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TemperatureSensor should render correctly 1`] = `
<div
  className="tempSensor"
  style={
    Object {
      "height": "920px",
      "width": "280px",
    }
  }
>
  <h4
    className="name"
  >
    Office
  </h4>
  <div
    className="value"
  >
    23.2
    Â°C
  </div>
</div>
`;
